src/redisearch_rs/inverted_index/src/lib.rs

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2006-Present, Redis Ltd.
+ * All rights reserved.
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0
+ * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+
+use std::io::{Read, Seek, Write};
+
+pub use ffi::{RSIndexResult, t_docId};
+
+/// Encoder to write a record into an index
+pub trait Encoder {
+    /// Write the record to the writer and return the number of bytes written. The delta is the
+    /// pre-computed difference between the current document ID and the last document ID written.
+    fn encode<W: Write + Seek>(
+        writer: W,
+        delta: t_docId,
+        record: &RSIndexResult,
+    ) -> std::io::Result<usize>;
+}
+
+pub enum DecoderResult {
+    /// The record was successfully decoded.
+    Record(RSIndexResult),
```

Comment:
Are we 100% positive that no data is borrowed from the index? It's all owned by the caller?

Reply:
To answer the question as to whether `RSIndexResult` will need lifetimes when written in Rust, the answer is no.

This is the C definition of `RSIndexResult`

https://github.com/RediSearch/RediSearch/blob/74451de0aac2c97bc6b684ae9b6b49c2fe00fb96/src/redisearch.h#L336-L374

The only pointers are `dmd` and `metrics`. However none of the encoders will these fields. So when we get to converting this to Rust we can make two new types:
1. Which contains the actual encoding / decoding data where everything is owned
2. Another type stored by the indexer. It will contain the base type in 1 with other metadata fields like metrics that might have lifetimes.

But this means whatever the encoders / decoders works with won't need lifetimes.